std/algebra/emulated/sw_bls12381/pairing.go

```diff
@@ -100,14 +100,122 @@ func (pr Pairing) Pair(P []*G1Affine, Q []*G2Affine) (*GTEl, error) {
 //
 // This function doesn't check that the inputs are in the correct subgroups.
 func (pr Pairing) PairingCheck(P []*G1Affine, Q []*G2Affine) error {
-	f, err := pr.MillerLoop(P, Q)
+	// check input size match
```

Comment:
The method seems to be almost fully inlined version of `millerLoopLInes` and `AssertFinalExpIsOne` (with some small differences). I would try to refer to those method directly. For example for `millerLoopLines` we can add boolean that we want to handle first loop iteration separately. Similarly for `AssertFinalExpIsOne`.

I think it is clearer and also more future-proof in case we have some other changes.